pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java

```diff
@@ -0,0 +1,247 @@
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
+        return send(record, null);
+    }
+
+    @Override
+    public Future<RecordMetadata> send(ProducerRecord<K, V> record, Callback callback) {
+        org.apache.pulsar.client.api.Producer producer;
+
+        try {
+            producer = producers.computeIfAbsent(record.topic(), topic -> createNewProducer(topic));
+        } catch (Exception e) {
+            callback.onCompletion(null, e);
+            CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
+            future.completeExceptionally(e);
+            return future;
+        }
+
+        Message msg = getMessage(record);
+
+        CompletableFuture<RecordMetadata> future = new CompletableFuture<>();
+        CompletableFuture<MessageId> sendFuture = producer.sendAsync(msg);
+        lastSendFuture.put(record.topic(), sendFuture);
+
+        sendFuture.thenAccept((messageId) -> {
+            future.complete(getRecordMetadata(record.topic(), msg, messageId));
+        }).exceptionally(ex -> {
+            future.completeExceptionally(ex);
+            return null;
+        });
+
+        future.handle((recordMetadata, exception) -> {
+            callback.onCompletion(recordMetadata, new Exception(exception));
+            return null;
+        });
+
+        return future;
+    }
+
+    @Override
+    public void flush() {
+        lastSendFuture.forEach((topic, future) -> {
+            try {
+                future.get();
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
+            }
+
+            // Remove the futures to remove eventually failed operations in order to trigger errors only once
+            lastSendFuture.remove(topic, future);
+        });
+    }
+
+    @Override
+    public List<PartitionInfo> partitionsFor(String topic) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Map<MetricName, ? extends Metric> metrics() {
+        return Collections.emptyMap();
+    }
+
+    @Override
+    public void close() {
+        close(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void close(long timeout, TimeUnit unit) {
+        try {
+            client.closeAsync().get(timeout, unit);
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private org.apache.pulsar.client.api.Producer createNewProducer(String topic) {
+        try {
+            return client.createProducer(topic);
+        } catch (PulsarClientException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Message getMessage(ProducerRecord<K, V> record) {
+        if (record.partition() != null) {
+            throw new UnsupportedOperationException("");
```

Comment:
does it mean this doesn't support kafka topic with partition?

Reply:
No, it's just that you cannot manually specify which partition to publish to. That would require some gimnastic to support. We could do it later.

Reply:
Ok.. I saw `PulsarKafkaConsumer` where we create individual `Consumer` for each partitioned. Similar way, here we could have created separate producer for each partition and then send message using key.

Reply:
Yes, that would be an option as well. I first did the producer part and used the regurlar partitioned producer thing, though it might make sense to deal with the partitions manuallly.

Reply:
Tried to do manual partitions management in the Producer end, but it gets complicated to deal with the Kafka `Partitioner` interface. The trick is that relies on a `Cluster` view which is overly Kafka specific. 

I'll leave it for later if there is a need to support these features too.